### DISCUSSION TITLE
Updating accel structure from GPU

### DISCUSSION THREAD

**Discussion by cat-state** [LuisaGroup/LuisaCompute#157] [Q&A] [answered]
Hi, is it possible to efficiently update the positions of procedural primitives in the accel structure from device side kernels? In the examples I only see building the structure up front once and then reusing it. My use-case is to reuse the fast ray-aabb tests for collision detection and I have a realtime gpu physics simulation id port to luisa if that were possible :)

**Comment by Mike-Leo-Smith** [comment, marked as answer]
You can launch a kernel to update the AABB buffer and then encode into the stream a ProceduralPrimitive::build() command and an Accel::build() command.

**Reply by cat-state** [reply]
oh ofc! thank you for the fast response :) 
before i try porting everything - do you know how many AABBs would be possible to update at 100Hz or so (on best hardware possible)?

**Reply by Mike-Leo-Smith** [reply]
I don’t have the specific data for now but I think several million won’t be a problem. I’ll run a quick test today or tomorrow.

**Reply by Mike-Leo-Smith** [reply]
I have run the tests on my 4060 laptop (70W TDP for GPU), where I randomly set the center of AABBs every frame and update the accel. For 100Hz update, if the accel is created with the `AccelOption::allow_update` flag on, the number is over 4M; if the option is off (so that the acceleration structure is forced to rebuild at every frame for better quality), the number is over 1M. I guess on better hardware like a desktop 4090, the numbers would further grow several times.
